include/bout/petsc_interface.hxx

```diff
@@ -0,0 +1,576 @@
+  PetscVector(const PetscVector<F> v, Vec vec) {
+#if CHECKLEVEL >= 2
+    int fsize, msize;
+    if (std::is_same<F, FieldPerp>::value) {
+      fsize = v.indexConverter->getMesh()->localSizePerp();
+    } else if (std::is_same<F, Field2D>::value) {
+      fsize = v.indexConverter->getMesh()->localSize2D();
+    } else if (std::is_same<F, Field3D>::value) {
+      fsize = v.indexConverter->getMesh()->localSize3D();
+    } else {
+      throw BoutException("PetscVector initialised for non-field type.");
+    }
+    VecGetSize(vec, &msize);
+    ASSERT2(fsize == msize);
+#endif
+    vector = vec;
+    indexConverter = v.indexConverter;
+    location = v.location;
+    initialised = true;
+}
+
+  ~PetscVector() {
+    // FIXME: Should I add a check to ensure the vector has actually been created in the first place? Is that possible in Petsc?
+    if (vector != nullptr && initialised) {
+      VecDestroy(&vector);
+    }
+  }
+
+  /// Copy assignment
+  PetscVector<F>& operator=(const PetscVector<F>& rhs) {
+    swap(*this, rhs);
+    return *this;
+  }
+
+  /// Move assignment
+  PetscVector<F>& operator=(PetscVector<F>&& rhs) {
+    vector = rhs.vector;
+    indexConverter = rhs.indexConverter;
+    location = rhs.location;
+    initialised = rhs.initialised;
+    rhs.vector = nullptr;
+    rhs.initialised = false;
+    return *this;
+  }
+
+  friend void swap<F>(PetscVector<F>& first, PetscVector<F>& second);
+
+  /// Assign from field, copying over field values. The vector must
+  /// already have been created for this to work.
+//  PetscVector<F>& operator=(F& rhs) {
+//    PetscInt ind;
+//    ASSERT1(initialised);
+//    ASSERT2(location == rhs.getLocation());
+//    BOUT_FOR(i, rhs.getRegion(RGN_ALL)) {
+//      ind = indexConverter->getGlobal(i);
+//      if (ind != -1) {
+//	VecSetValues(vector, 1, &ind, &rhs[i], INSERT_VALUES);
+//      }
+//    }
+//    assemble();
+//    return *this;
+//  }
+
+  PetscVectorElement& operator()(ind_type& index) {
+#if CHECKLEVEL >= 1
+    if (!initialised) {
+      throw BoutException("Can not return element of uninitialised vector");
+    } 
+#endif
+    int global = indexConverter->getGlobal(index);
+#if CHECKLEVEL >= 1
+    if (global == -1) {
+      throw BoutException("Request to return invalid vector element");
+    }
+#endif
+    return PetscVectorElement::newElement(&vector, global);
+  }
+  
+  void assemble() {
+    VecAssemblyBegin(vector);
+    VecAssemblyEnd(vector);
+  }
+  
+  void destroy() {
+    if (vector != nullptr && initialised) {
+      VecDestroy(&vector);
+      vector = nullptr;
+      initialised = false;
+    }
+  }
+
+  /// Returns a field constructed from the contents of this vector
+  const F toField() {
+    F result(indexConverter->getMesh());
+    result.allocate();
+    result.setLocation(location);
+    PetscScalar val;
+    PetscInt ind;
+    // Note that this only works when yguards have a width of 1.
```

Comment:
In principle yes. However, the more I've thought about this the more I've realised it's a bit tricky. The issue is that it seems  only one guard cell will be used to enforce boundary conditions, even when the width of the y-guard is 2 (at least, assuming I'm understanding things correctly). So, it might be that a width of 1 is actually what we want and i should change the x-boundaries accordingly. However, I'm not confident on this matter, so feel free to correct me.

Reply:
My feeling is that practically we can/should assume that PETSc operators only use one guard-cell or boundary-cell. As far as I know we don't know any good way to impose higher-order boundary conditions that require 2 (or more) points, so any differential operators that we're putting in a PETSc matrix are essentially restricted to 3-point stencils (in a single dimension) anyway. I wouldn't worry about generalising to more than 1 guard/boundary cell unless/until someone actually wants to implement an operator that requires it (I suspect this will never happen).

\* You might reasonably ask what that means for the things we do that do require two boundary cells... One case is upwinding derivatives like `U2` or `W3` that use a wider-than-3-point stencil, but presumably any flow is into the boundaries, so the stencil will always be biased away from the boundary cells. Another case is `interp_to`, which uses values from both boundary cells; I don't have any justification for this case, beyond that the simulations seem to run...

Reply:
I've ensured that only one layer of boundary cells will be included and added some comments to this effect.